plane/src/typed_unix_socket/server.rs

```diff
@@ -0,0 +1,263 @@
+                                        stream,
+                                        event_tx.clone(),
+                                        request_tx.clone(),
+                                        response_tx.subscribe(),
+                                        shutdown_handler_tx.subscribe(),
+                                    ));
+                                }
+                                Err(e) => {
+                                    tracing::error!("Error accepting connection: {}", e);
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        });
+
+        Ok(Self {
+            socket_path,
+            event_tx,
+            request_tx,
+            response_tx,
+            shutdown_server_tx,
+            shutdown_handler_tx,
+        })
+    }
+
+    pub fn subscribe_events(&self) -> broadcast::Receiver<MessageToServer> {
+        self.event_tx.subscribe()
+    }
+
+    pub fn subscribe_requests(&self) -> broadcast::Receiver<WrappedMessage<MessageToServer>> {
+        self.request_tx.subscribe()
+    }
+
+    pub async fn send_response(
+        &self,
+        request: &WrappedMessage<MessageToServer>,
+        response: MessageToClient,
+    ) -> Result<(), Error> {
+        // Wait until there is at least one subscriber
+        let mut backoff = get_quick_backoff();
+        while self.response_tx.receiver_count() == 0 {
+            tracing::info!("Waiting for a subscriber...");
+            backoff.wait().await;
+        }
+
+        let response = WrappedMessage {
+            id: request.id.clone(),
+            message: response,
+        };
+
+        self.response_tx.send(response)?;
+        Ok(())
+    }
+
+    pub async fn send_message(&self, message: MessageToClient) -> Result<(), Error> {
+        let message_msg = WrappedMessage { id: None, message };
+
+        // Wait until there is at least one subscriber
+        let mut backoff = get_quick_backoff();
+        while self.response_tx.receiver_count() == 0 {
+            tracing::info!("Waiting for a subscriber...");
+            backoff.wait().await;
+        }
+
+        self.response_tx.send(message_msg)?;
+        Ok(())
+    }
+}
+
+impl<MessageToServer, MessageToClient> Drop
+    for TypedUnixSocketServer<MessageToServer, MessageToClient>
+where
+    MessageToServer: Send + Sync + 'static + Clone + Debug + Serialize + for<'de> Deserialize<'de>,
+    MessageToClient: Send + Sync + 'static + Clone + Debug + Serialize + for<'de> Deserialize<'de>,
+{
+    fn drop(&mut self) {
+        // Signal the server loop to shut down
+        let _ = self.shutdown_server_tx.send(());
+
+        // Signal the handler to shut down
+        let _ = self.shutdown_handler_tx.send(());
+
+        if let Err(e) = fs::remove_file(&self.socket_path) {
+            tracing::warn!("Failed to remove socket file: {}", e);
+        }
+    }
+}
```

Comment:
The drop/clone semantics here are inconsistent, creating the footgun you encountered while testing. In general I would expect clone/drop to behave such that the code in the block is essentially a no-op as far as `a` is concerned:

```rust
let a = Something::new();

{
    let b = a.clone()
}
```